[SPARK-6406] Launch Spark using assembly jar instead of a separate launcher jar

nishkamravi2 opened the pull request.

nishkamravi2 committed:
Fix for SPARK-1758: failing test org.apache.spark.JavaAPISuite.wholeTextFiles

The prefix "file:" is missing in the string inserted as key in HashMap

nishkamravi2 committed:
Fix in Spark for the Concurrent thread modification issue (SPARK-1097, HADOOP-10456)

nishkamravi2 committed:
Undo the fix for SPARK-1758 (the problem is fixed)

nishkamravi2 committed:
Improved fix for ConcurrentModificationIssue (Spark-1097, Hadoop-10456)

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark

nishkamravi2 committed:
Accept memory input as "30g", "512M" instead of an int value, to be consistent with rest of Spark

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

nishkamravi2 committed:
Modify default YARN memory_overhead-- from an additive constant to a multiplier (redone to resolve merge conflicts)

nishkamravi2 committed:
Update YarnAllocator.scala

nishkamravi2 committed:
Improving logging for AM memoryOverhead

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

Conflicts:
	yarn/common/src/main/scala/org/apache/spark/deploy/yarn/ClientBase.scala

nishkamravi2 committed:
Additional changes for yarn memory overhead

nishkamravi2 committed:
Additional changes for yarn memory overhead issue

nishkamravi2 committed:
Additional documentation for yarn memory overhead issue

nishkamravi2 committed:
Remove out

nishkamravi2 committed:
Slight change in the doc for yarn memory overhead

nishkamravi2 committed:
Doc change for yarn memory overhead

nishkamravi2 committed:
Update YarnAllocator.scala

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

Conflicts:
	yarn/common/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala
	yarn/common/src/main/scala/org/apache/spark/deploy/yarn/ClientBase.scala
	yarn/common/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

nishkamravi2 committed:
Changes to maxResultSize code (improve error message and add condition to check if maxResultSize > 0)

nishkamravi2 committed:
Modify test for maxResultSize

nishkamravi2 committed:
Update TaskSetManager.scala

nishkamravi2 committed:
Update TaskSetManagerSuite.scala

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

nishkamravi2 committed:
Workaround for IllegalStateException caused by recent changes to BlockManager.stop

nishkamravi2 committed:
Update DiskBlockManager.scala

nishkamravi2 committed:
Update DiskBlockManager.scala

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

nishkamravi2 committed:
Try-catch block around the two occurrences of removeShutDownHook. Deletion of semi-redundant occurrences of expensive operation inShutDown.

nishkamravi2 committed:
Merge branch 'master_nravi' of https://github.com/nishkamravi2/spark into master_nravi

Conflicts:
	core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

nishkamravi2 committed:
Update HadoopRDD.scala

nishkamravi2 committed:
Update NewHadoopRDD.scala

nishkamravi2 committed:
Update NewHadoopRDD.scala

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

nishkamravi2 committed:
"hadoop" string in the assembly name should not be mandatory (everywhere else in spark we mandate spark-assembly*hadoop*.jar)

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java:
```diff
@@ -341,7 +341,7 @@ private String findAssembly(String scalaVersion) {
     }
 
-    final Pattern re = Pattern.compile("spark-assembly.*hadoop.*\\.jar");
+    final Pattern re = Pattern.compile("spark-assembly.*[hadoop]*.*\\.jar");
```
Shouldn't this bit be `.*(hadoop)?.*` but then can't it just collapse to matching anything with `.*`?

nishkamravi2 committed:
Update AbstractCommandBuilder.java

nishkamravi2 commented:
Thanks @srowen, that looks better.

nishkamravi2 committed:
Update AbstractCommandBuilder.java

vanzin commented:
Wait, I'm confused. What's the bug? Even your own comment says that the "hadoop" part is mandatory elsewhere. Here's the original shell script code:

https://github.com/apache/spark/blob/branch-1.3/bin/compute-classpath.sh#L79

"hadoop" is definitely mandatory there.

nishkamravi2 commented:
Please ignore the comment above (I misread the regex). However, we do need to relax the check on "hadoop". CDH itself names the outermost jar spark-assembly.jar. As to why we did not catch this issue with compute-classpath.sh, short answer: because we had our own custom version of it.

nishkamravi2 commented:
And btw, we need to check this in

nishkamravi2 commented:
Sorry, clicked on the close button in error.

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

vanzin commented:
Nishkam, can you explain why we need to remove "hadoop" from the regex now when the previous code had it? Can you explain one use case where the current code fails and the previous code worked?

CDH ships the correctly named assembly with a symlink named "spark-assembly.jar". So it works fine with the existing code.

nishkamravi2 commented:
Marcelo, consider the following use case and see if it makes sense: I'm a savvy customer. I name my jar file "spark-assembly.jar" because I see that as the outermost jar file in CDH (I get rid of unnecessary symlinks). I modify (or rather bypass) most of compute-classpath.sh including the check for jar file name. Now I find that the check for the jar file name (which was a bad idea to begin with) is part of Spark code that gets compiled and therefore cannot be easily modified. 

Btw, is having strict checks on assembly jar names a good idea? Or should I be able to name my jar "assembly.jar" as long as it is the only one and in the right folder?

nishkamravi2 committed:
Update spark-class

nishkamravi2 commented:
Also, we should not look for a separate launcher library jar. Like any other spark component it should be part of the uber jar. Modified PR accordingly (please see modified spark-class).

srowen commented:
Going "off road" is up to the adventurous consumer. I think I'd produce a build with standard `make-distribution.sh` tools or the like, which should dump out a binary with `hadoop` in the name anyway, and link to that. I don't think I would modify scripts anyway.

That said, yeah, why not just look for `spark-assembly.*\.jar`? I don't see why `hadoop` is so important there. That's a plausible way to identify _the_ intended assembly JAR. Is there an outside risk that somebody is running with two `spark-assembly` files in the same dir, one of which has `hadoop` in the name, and this breaks that?

My hunch is there's a reason for the separate JAR, like decoupling the launcher from the launchee as was done previously with the scripts.

Anyway, net-net at the moment I'd support loosening the regex but it is a sort of theoretical problem.

vanzin commented:
Please *DO NOT* make a change to not have a separate launcher jar. That choice was explicit, because it's much, MUCH faster to load a jar file that is a few kB large than loading the uber jar.

Seriously, I still don't see any bug here.

vanzin commented:
>  I'm a savvy customer. I name my jar file "spark-assembly.jar" because I see that as the outermost jar file in CDH (I get rid of unnecessary symlinks).

Well, in my view the customer is on his own as soon as he starts to modify the distribution. If he makes that change in 1.3, he'll have to figure out he needs to change other scripts too. Yes, it's easier to change a shell script than it is to change a compiled library, but I still don't see why Spark needs to support a random user who wants to deploy his own view of what the Spark distribution should look like. If that user is really savvy, he's savvy enough to fix what needs to be fixed on his own.

The problem with looseing the regex is theoretical, as Sean says. But consider this output, after running `sbt package assembly`:

    $ ls -l assembly/target/scala-2.10/
    total 156928
    -rw-rw-r-- 1 vanzin vanzin 160683349 Mar 18 06:50 spark-assembly-1.3.0-SNAPSHOT-hadoop2.5.0.jar
    -rw-rw-r-- 1 vanzin vanzin       290 Mar 20 09:35 spark-assembly_2.10-1.4.0-SNAPSHOT.jar

Now you have a non-theoretical breakage because of your change.

nishkamravi2 commented:
Are we never loading the uber jar in this implementation? Can you quantify the gains you think you get by loading a separate launcher jar as compared to the uber jar? Are we convinced that those are significant enough to take away the simplicity of a single uber jar and potentially breaking backward compatibility? An uber jar allows a customer to copy it anywhere and point their classpath to it. 

With regards to the assembly jar name issue: I'm not sure the above example is a breakage if it can be fixed by simply deleting one of those two jar files once the Spark launcher prints an explicit message stating so. In general, I'm not convinced that a strict assembly name check should be a solution for anything. But, if you are, a good compromise would be to do that check in spark-submit (or an outer shell script) than hard-coding it in the Spark code. 

Sean, I think from a customer's point of view, the launcher is the application and the launchee is Spark library. That simplicity is desired, as compared to separating out stub code within Spark into a different library (for very little in return) and increasing complexity?

vanzin commented:
What backwards compatibility concerns do you have? The library didn't exist before, so it's pretty hard to think of any.

As for performance, I didn't measure explicitly, but even if the gain is not that big, I really don't see why you see this as complicating anything.

The issue with the breakage is not that it can be easily fixed. Is that it would exist at all. You'd be forcing people to fix their build when they did absolutely nothing wrong. What I haven't see here is any argument that supports relaxing the regex without using some use case where the user is intentionally breaking the Spark distribution.

vanzin commented:
And since you asked, here are times for the following command, over 100 runs:

    SPARK_HOME=$PWD java -cp $JAR org.apache.spark.launcher.Main org.apache.spark.deploy.SparkSubmit

Against the assembly:

    avg: 0.156157615185, min: 0.148868083954, max: 0.186276912689

Against the launcher jar:

    avg: 0.101513769627, min: 0.0895018577576, max: 0.130180120468

So it's about 30% faster on average (with an SSD).

nishkamravi2 commented:
That's not a 30% gain. That's a constant gain of 0.05 seconds.

> What backwards compatibility concerns do you have? The library didn't exist before, so it's pretty hard to think of any.

The library didn't exist before and now it's required. That's a backward compatibility issue.

> As for performance, I didn't measure explicitly, but even if the gain is not that big, I really don't see why you see this as complicating anything. 

Not sure what it brings to the table then?

> The issue with the breakage is not that it can be easily fixed. Is that it would exist at all. You'd be forcing people to fix their build when they did absolutely nothing wrong. What I haven't seen here is any argument that supports relaxing the regex without using some use case where the user is intentionally breaking the Spark distribution.

Your arguments may be considered somewhat contradictory to idea of this launcher library which is to provide additional flexibility in deploying Spark. For every customer who wants to launch Spark programmatically (which itself is a corner case), there is one who has a custom Spark deployment setup.  

Besides, as I said, if you are convinced that strict checking of assembly file name is a good idea (which is very subjective), it is better to do that in a shell script than hardcode it in Spark code that gets compiled. 

I think it may be best to weigh in opinion from others at this point @srowen @pwendell @andrewor14

vanzin commented:
> The library didn't exist before and now it's required. That's a backward compatibility issue.

That is not a backwards compatibility issue. If you try to launch your app against Spark 1.2 or 1.3 it will still work. If it didn't, *that* would be a backwards compatibility issue. The library being there is completely internal to Spark, and no user should need to care about it being there.

>  there is one who has a custom Spark deployment setup. 

As I've said before, if someone is customizing the layout of the Spark installation, it's up to them to fix what needs to be fixed. The shell scripts before made the same assumptions the launcher library currently does. So nothing has been broken.

nishkamravi2 commented:
1. What do you gain by creating a separate launcher library? If you deviate from what we had before you need a compelling argument, not the other way round.
2. If at all needed, is it better to do a spark assembly file name check in spark-submit (or some such) than hardcode it in Spark?

vanzin commented:
re: 1, you gain 50ms (out of 150ms) on every invocation of the library. Might not be much, but there *is* a difference.

re: 2, why is it better? You haven't yet described a single case where it is better. It shouldn't matter for anyone, except someone who's trying to create a different layout for the Spark distribution, and I don't see any compelling argument for supporting that use case.

vanzin commented:
On top of that, placing the logic to find the assembly in a shell script means that now `CommandUtils.scala` would need to either replicate that logic in Scala, or execute the shell script (which is what it did before).

srowen commented:
There are two questions here.

"Should assembly file names be required to contain `hadoop`?"

Yes argument: it's the current behavior and might avoid picking up other random assembly-like JARs
No argument: might help people pick up a customized assembly JAR more naturally.

I personally think it doesn't matter much, and all else equal would leave it, considering the discussion here. Is there an end-user problem that hasn't been voiced yet?

"Should the launcher .jar be separate from the assembly?"

Yes argument: slightly faster startup. Maybe a desirable decoupling? it's how it works now.
No argument: deployment now means managing two files

I personally don't think the startup time difference matters. Right? This is a deployment question. It's not worse than when one had to ship around both scripts and an assembly, so I find this the same as before. That doesn't mean there isn't an argument for one uber jar.

Is that it then? question is: should the launcher be a part of the uber jar to simplify deployment over the current and past state of things, at the cost of a little startup time hit? I could see that one go either way.

vanzin commented:
>  Is there an end-user problem that hasn't been voiced yet?

No, but there is an end-user issue that has been voice already: it can break launching jobs after you build spark with sbt (e.g. `sbt package assembly`). That's why the old scripts also had the same naming restriction.

nishkamravi2 commented:
Marcelo and I discussed this offline and we have agreed on how to resolve this issue:

1. We will remove the check for launcher library and look for the uber jar instead
2. For the regex expression problem, we will move that check out into spark-class (with hadoop string included in the check) and pass the assembly file location as an argument to Launcher. That way we can completely eliminate the check for spark assembly jar file name from Spark code.

Will modify the PR accordingly after some testing

vanzin commented:
I talked to @nishkamravi2 offline, and we agreed on a different approach: getting rid of the launcher library (the perf difference not being in any way worrisome), and passing the location of the assembly file to the launcher's `Main` class.

That way the logic to find the assembly (still with the `*hadoop*` restriction) replaces the logic to find the launcher in the scripts, and the launcher library doesn't need code to look for the assembly, which should make use cases similar to his easier to handle.

nishkamravi2 committed:
spark-class changes

nishkamravi2 committed:
spark-class merge
Merge branch 'master_nravi' of https://github.com/nishkamravi2/spark into master_nravi

Conflicts:
	bin/spark-class

pwendell commented:
Hey @vanzin - can you explain a bit why the performance impact isn't worrisome? I think a first order goal was to avoid invoking multiple JVM's that have the assembly jar in the classpath, because it is slow.

vanzin commented:
> can you explain a bit why the performance impact isn't worrisome?

Because it doesn't really save a significant amount of time. Invoking `org.apache.spark.launcher.Main` against the small library and the uber jar results in ~50ms extra for the uber jar in my system (probably a little more if you have spinning disks). That would still make the new launch path much faster than the old one (see the measurements I did in the original bug, SPARK-4924).

pwendell commented:
What is the intuitive explanation for why it is faster than before? It seems like launching the assembly twice would be slower.

Separately, is it worth doing a big refactoring for this issue? If we want to support this maybe we should just widen the regex. Or do you now feel the refactoring is an important design improvement, and it was simply instigated by this issue being reported?

For instance, if a change makes starting spark-shell > 100ms slower for people, IMO that's bad usability. Responsiveness for things like this can matter when people getting an impression of the overall performance of a system, and we have components specifically designed for quick interactive experience.

nishkamravi2 commented:
Hey @pwendell, a constant sub-second increase in execution time during the lifetime of an app should be acceptable. We tend to regress significantly more than with every Spark version as we pile up more code (we generally try to avoid reporting regressions of < 5%). Having to worry about two jar files as opposed to one on the other hand seems like a much more fundamental usability concern IMO.

pwendell commented:
@nishkamravi2 - It's about the startup time of the interactive shell. That's the usability trade off. If the shell takes 300 extra ms to start up, it is not acceptable from a usability perspective. This is not about performance for a long running batch processing job.

vanzin commented:
When I said "before" I meant the code currently in branch-1.3. It's faster now, even with using the assembly for both invocations, because that's still loading the assembly less times than it did before (which could be up to 3 IIRC), it avoids invoking `java` just to see what version it is, and Java code is generally faster than bash code.

The added overhead of using the assembly for the first invocation (instead of the smaller jar) is not enough to make performance be as slow as before; it's still considerably faster (400ms before vs 150ms now in my system). So that's not really a big source of worries here, and the code changes shouldn't really be too bad.

Also, I'd argue that most of the startup time of spark-shell is actually loading all the Scala classes (which the launcher invocation doesn't do since it's pure java) and contacting the cluster manager. That should make the overhead of the startup script be mostly insignificant.

The remainder change then just builds on top of this change; instead of having code both in the scripts and in the library to find the assembly, it's just easier to tell the library where the assembly is (since the scripts already know).

vanzin commented:
BTW just a pre-emptive comment that if you really feel like going down this path, you'll have to move the "invalid CEN header" (SPARK-1703) checks from AbstractCommandBuilder.java back to the shell scripts.

pwendell commented:
Okay - if there is no big performance hit in launching from 1.3->1.4, that's fine with me.

I guess then the remaining question is whether to make a smaller change (simply make the wildcard match more liberal) vs. doing a larger change.

vanzin commented:
Loosening the wildcard match by itself can break sbt builds, so that doesn't sound like a good plan unless the wildcard still avoids matching other sbt-generated artifacts.

vanzin commented:
Just to clarify, personally, I still thing there's nothing wrong and that anybody who uses Spark the way it expects to be used (i.e. make-distribution.sh + calling bin/spark-submit) is unaffected by any recent changes. Nishkam feels strongly that I've broken his use case, even though I believe his use case is technically unsupported since he's modifying the launch scripts. But if he can change the library in a way that achieves many of the same original goals and still supports his use case, then I'm fine with it.

Pulling more logic into the shell scripts is something that goes against those goals, but it feels like a small enough change that I'm willing to overlook it.

nishkamravi2 commented:
Sorry, I'm out for the weekend, not able to respond to the comments on this PR as promptly as I'd like to. I can make the changes and update the PR on Monday (hopefully) after some testing. If at that point we feel that the PR is not helping, we can certainly choose to not commit it. I think that the changes are small and in the right direction.

nishkamravi2 committed:
Launcher library changes (SPARK-6406)

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

nishkamravi2 commented:
@vanzin @pwendell @srowen  PR updated with the changes discussed last week. Also, added a check for JAVA_HOME (please see if this makes sense). Trying to keep the changeset small and focussed.

vanzin commented on bin/spark-class:
```diff
@@ -40,36 +40,24 @@ else
-  SPARK_LAUNCHER_CP="$LAUNCHER_DIR/classes"
+# Find assembly jar
+SPARK_ASSEMBLY_JAR=
+ASSEMBLY_DIR="$SPARK_HOME/lib"
```
Where are you looking for the assembly under `assembly/target/scala-$SPARK_SCALA_VERSION`?

That's needed to not break dev builds.

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/Main.java:
```diff
@@ -47,10 +47,14 @@
    * script.
    */
+
+  static String uberJarPath;
```
This looks really ugly. I'd really prefer plumbing this to the command builders through the constructor. It's a little bit more code but much cleaner.

vanzin commented:
As I mentioned before, you need to take the "invalid CEN header" check from AbstractCommandBuilder and do it in the shell scripts now, otherwise that check is useless.

nishkamravi2 commented:
Thanks for the comments @vanzin. Will address them soon.

nishkamravi2 committed:
Changes for SPARK-6406

nishkamravi2 commented:
Addressed two of the three comments @vanzin. The third one is tricky, let's discuss tomorrow when you're around.

nishkamravi2 committed:
Locate the jar using getLocation, obviates the need to pass assembly path as an argument

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

Conflicts:
	launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java

nishkamravi2 commented:
PR updated (instead of passing assembly jar on cmdline, we can find it dynamically using getLocation).

vanzin commented on bin/spark-class:
```diff
@@ -40,35 +40,44 @@ else
+  JAR_CMD="jar"
+fi
+
+jar_error_check=$("$JAR_CMD" -tf "$SPARK_ASSEMBLY_JAR" nonexistent/class/path 2>&1)
```
You may need to add something like #4873 here.

vanzin commented on bin/spark-class:
```diff
@@ -40,35 +40,44 @@ else
-SPARK_LAUNCHER_CP=
+# Find assembly jar
+SPARK_ASSEMBLY_JAR=
 if [ -f $SPARK_HOME/RELEASE ]; then
```
You're not changing this, but could you add quotes here?

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java:
```diff
@@ -182,7 +186,7 @@ void addOptionString(List<String> cmd, String options) {
     }
 
-    String assembly = findAssembly();
+    final String assembly = AbstractCommandBuilder.class.getProtectionDomain().getCodeSource().getLocation().getPath();
```
Can you do just `getClass()` here instead of the more verbose version? Also, line looks pretty long.

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/Main.java:
```diff
@@ -31,7 +31,7 @@
 
   /**
-   * Usage: Main [class] [class args]
+   * Usage: Main [uberJarPath] [class] [class args]
```
Changes to this file are not needed anymore, right?

vanzin commented:
Can an admin trigger jenkins for this PR please?

@nishkamravi2 , I think you can also remove the line in make-distribution.sh that packages the launcher jar. Otherwise, some minor things, but looks good.

Can you also change the PR title since it's not accurate? Something like "Launch Spark using the assembly instead of a separate launcher jar".

andrewor14 commented:
retest this please

SparkQA commented:
[Test build #29144 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29144/consoleFull) for   PR 5085 at commit [`4a45d6a`](https://github.com/apache/spark/commit/4a45d6adc12ac24df284aa2dc65fe915ad11243f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29144 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29144/consoleFull) for   PR 5085 at commit [`4a45d6a`](https://github.com/apache/spark/commit/4a45d6adc12ac24df284aa2dc65fe915ad11243f).
 * This patch **fails Spark unit tests**.

 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `    implicit class DslSymbol(sym: Symbol) extends ImplicitAttribute `
  * `case class Alias(child: Expression, name: String)(`
  * `class Column(protected[sql] val expr: Expression) extends Logging `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29144/
Test FAILed.

vanzin commented:
I can't comment on random code that you didn't change, but you'll need to change the code around this code in AbstractCommandBuilder:

     needsDataNucleus = assemblyJar.getEntry("org/apache/hadoop/hive/ql/exec/") != null;

You can just skip the check for Hive and add all datanucleus jars if they're present. That's the same thing Sean did in the PR I mentioned before. (Bonus points for getting rid of the annoying message printed to stderr.)

nishkamravi2 committed:
Merge branch 'master' of https://github.com/apache/spark into master_nravi

nishkamravi2 committed:
Minor changes and bug fixes

nishkamravi2 committed:
Update Main.java

nishkamravi2 commented:
Addressed comments, fixed a couple of bugs. Can this be retested? Thanks.

srowen commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #29295 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29295/consoleFull) for   PR 5085 at commit [`bad4349`](https://github.com/apache/spark/commit/bad43499ed2518e0d5153ac56ce03beca3980e2c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29295 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29295/consoleFull) for   PR 5085 at commit [`bad4349`](https://github.com/apache/spark/commit/bad43499ed2518e0d5153ac56ce03beca3980e2c).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29295/
Test PASSed.

vanzin commented:
LGTM.

srowen commented:
Given @vanzin and @nishkamravi2 have thoroughly discussed and approved this, I'm inclined to merge it. The changes look sane and do simplify the launcher somewhat.

andrewor14 commented:
@vanzin @nishkamravi2 When I try to run `bin/spark-shell --master local-cluster[2,1,512]` my executors keep failing complaining that the scala classes are not found. I dug into this a little and it turns out that the assembly jar path from [this line](https://github.com/apache/spark/blob/d7f2c198678d1f5ffc1c6da3a6fb189b0a4ef070/launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java#L189) is not what I expected: `$SPARK_HOME/launcher/target/scala-2.10/classes/`. Before this commit, it used to be the path of the assembly jar itself.

Is this a known issue?

andrewor14 commented:
Looks like @davies filed a JIRA for this already: https://issues.apache.org/jira/browse/SPARK-6890

nishkamravi2 commented:
Hey @andrewor14, is this a Mac/Windows issue? I can't seem to be able to reproduce it on Linux. Also, do you happen to have the path that was returned by getLocation().getPath() ?

andrewor14 commented:
Yes, the path was `$SPARK_HOME/launcher/target/scala-2.10/classes/`. I'm in a mac.
